### DISCUSSION TITLE
Add multiple "to"s in transitions

### DISCUSSION THREAD

**Discussion by HamzaFarhan** [DAGWorks-Inc/burr#339] [Help + Q&A] [answered]
I noticed that the "from" in a transition can be a list of strings. So multiple actions can go to the same action if a certain condition is met:
(["action_a", "action_b"], "action_c", CONDITION)
What about going from one action to multiple actions if a condition is met?
("action_a", ["action_b", "action_c"], CONDITION)
Would updating the with_transitions function in GraphBuilder be enough?
Or would this lead to some sequencing issues? Or maybe it's our responsibility as devs to not go to multiple actions that rely on each other.

```python
def with_transitions(
        self,
        *transitions: Union[
            Tuple[Union[str, list[str]], Union[str, list[str]]],
            Tuple[Union[str, list[str]], Union[str, list[str]], Condition],
        ],
    ) -> "GraphBuilder":
        """Adds transitions to the graph. Transitions are specified as tuples of either:
            1. (from, to, condition)
            2. (from, to)  -- condition is set to DEFAULT (which is a fallback)

        Transitions will be evaluated in order of specification -- if one is met, the others will not be evaluated.
        Note that one transition can be terminal -- the system doesn't have

        :param transitions: Transitions to add
        :return: The application builder for future chaining.
        """
        if self.transitions is None:
            self.transitions = []
        for transition in transitions:
            from_, to_, *conditions = transition
            if len(conditions) > 0:
                condition = conditions[0]
            else:
                condition = default
            if not isinstance(from_, list):
                from_ = [from_]
            if not isinstance(to_, list):
                to_ = [to_]
            for action_from in from_:
                for action_to in to_:
                    if not isinstance(action_from, str):
                        raise ValueError(
                            f"Transition source must be a string, not {action_from}"
                        )
                    if not isinstance(action_to, str):
                        raise ValueError(
                            f"Transition target must be a string, not {action_to}"
                        )
                    self.transitions.append((action_from, action_to, condition))
        return self
```

**Comment by skrawcz** [comment]
~I think that should be enough.~ @elijahbenizzy right?

[post script] after Elijah's comment below, yeah what's the context that this would make sense for? [/post script]

**Reply by elijahbenizzy** [reply]
Taking a step back -- I'm not sure this makes sense? In what case would we transition from one to multiple? 

`a -> b given FOO`
`a -> c given FOO`

Would always select `a->b`, and never `a -> c`. If we had multi-edges and multiple concurrent states (something we're thinking through) it would be useful...

**Reply by elijahbenizzy** [reply]
Oh is this what you meant?

It could also be that it produces a list of actions, but I think that's more about structuring the DAG.

E.G. if it's `a -> b, c`, then we want `a -> b -> c` (effectively replacing `a->b`, `b->c`

That makes sense to me.

**Reply by HamzaFarhan** [reply]
I want certain actions to loop until I'm satisfied with the result. But with a limit on the number of attempts.
```python
[
    ("answer_question", "evaluate_answer"),
    # evaluate the answer and increment action_attempts if the evaluation is not satisfied and action_attempts is less than the limit
    (
        "evaluate_answer",
        ["increment_action_attempts", "answer_question"],
        ~expr("answer_evaluation.evaluation")
        & expr(f"action_attempts < {ACTION_ATTEMPTS_LIMIT}"),
    ),
    # if the evaluation is not satisfied and action_attempts is greater than or equal to the limit, reset action_attempts and go to the error_message action
    (
        "evaluate_answer",
        ["reset_action_attempts", "error_message"],
        ~expr("answer_evaluation.evaluation")
        & expr(f"action_attempts >= {ACTION_ATTEMPTS_LIMIT}"),
    ),
    # if the evaluation is satisfied, reset action_attempts and go to the final_message action
    (
        "evaluate_answer",
        ["reset_action_attempts", "final_message"],
        expr("answer_evaluation.evaluation"),
    ),
]
```

I can't go from `evaluate_answer` to `increment_action_attempts` and then from `increment_action_attempts` to the next action because `increment_action_attempts` would not know which action is being looped.

For now, I just increment or reset `action_attempts` at the end of evaluate_answer. But a standalone action would be better for reusability.

**Reply by HamzaFarhan** [reply]
Basically, how would I do this?

![Screenshot from 2024-08-28 11-37-05](https://github.com/user-attachments/assets/e749865f-ec2b-459b-9fb6-fa0c1f13eff9)

**Reply by HamzaFarhan** [reply]
I don't even need `ACTION D`. Just to go from `A` to `B` *AND* `C`.

**Reply by elijahbenizzy** [reply, marked as answer]
OK. So the way I'd model it is, given:

1. An input node (`get_question`, `A`)
2. An answering node (`answer_question`, `B`)
3. An evaluation node (`evaluate_answer`, `C`)
4. A final node (`final`, `D`)

`A->B` (always try to answer first)
`B->C` (then always evaluate the answer
`C->B`, if attempts < max_attempts (make sure `C` increments attempts as well, or stores a history of actions)
`C->D` if attempts >= max_attempts

Note that you could also go `A->B`, `B->C`, and `C->A`, while also going `A->D` if ready. I'm not 100% sure this captures your case though... 

My issue with the phrasing of "AND" is it's really to a chain of them (which is fair), but `AND` could mean parallel, in series, etc... And it's not clear from this.

One other option to think about:

```
.with_actions(B_C=chain("B", "C"), ...)
.with_transitions(
  ("A", "B_C"),
  ("B_C", "B_C", expr(...))
  ("B_C", "D", ...)
)
```

**Reply by HamzaFarhan** [reply]
My current model is exactly the one that you have proposed.
I initially wanted a separate incrementer node instead of "make sure C increments attempts as well".
But yesterday I had to add some extra "action attempts logic" inside `C`. So having custom incrementing logic inside custom evaluator nodes seems like a better idea instead of a dedicated `increment_attempts` node.
I still think triggering `n` **parallel** actions from a single action could be useful. But not something I need right now.

**Reply by yiweig** [reply]
Hello! @elijahbenizzy (apologies if direct tagging is not considered good form here) I came across this discussion post while exploring burr for our use case. I have a similar question as OP, where I would explicitly like to define multiple transitions from `A` to both `B` and `C` -- except my use case would benefit from explicitly defining that `B` and `C` happen in **parallel**.

Is this currently possible? Looking at this discussion it doesn't seem so, but I want to make sure since OP's use case was solved a different way.

Happy to open this as a separate discussion if appropriate. Thanks!

**Reply by elijahbenizzy** [reply]
Hey @yiweig, direct tagging is OK. So this is all doable now, but we're actively making it easier. Currently, parallelism is done with recursive subgraphs. You launch `n` subgraphs using your own executor, and it's easy to wire through the tracker.

See the docs [here](https://burr.dagworks.io/concepts/recursion/).

We're currently building out the notion of a MapReduceAction that'll take state and map it to actions/subgraphs, then take the results and merge them back in the state. It'll make the above a lot simpler to do for the parallel use-case!

Planning this next week, probably release the week after (TBD, still scoping), but if you want to get started I'd run it in parallel (treating it as a sub-application/sub-graph). Going to open up a PR with a proposed API, and would love your feedback!

**Reply by skrawcz** [reply]
@yiweig to add to @elijahbenizzy's current approach to do it. The mental model is:

1. Define an action that uses something like job lib / multi-processing/ async gather /etc  to do things in parallel.
2. To see it wired through in the UI you currently have to define a "Sub Application" (otherwise open telemetry should just work if it's an API call).

Your edge then is to that action that underneath does things in parallel. It is then responsible for using the right state and setting it back.

If it's two API calls you don't need the sub application part I don't think -- likely over kill; can you provide more info on what you want to make parallel? :) Otherwise we're working on making some of the ergonomics to make the simpler case easier -- and understanding use cases like yours would help.

**Reply by yiweig** [reply]
Awesome, thanks for the guidance guys! I actually did come across the Recursive Applications page on my initial reading of the docs, but I skimmed it too quickly and didn't realize it was talking about parallelism 😅 

@skrawcz sure! It's actually very similar to the example that @elijahbenizzy provided: we have multiple objects in our State and we want to call an LLM for each of them with a prompt. Since the LLM calls are quite slow (we're using OpenAI's chat completion) we wanted to dispatch those in parallel.

Admittedly, this can definitely be done like you described using a job lib / multi-processing / etc in a single Action. But I reached out because I wanted to see if there was a way to do it in `burr`.

Excited to see the PR! We'll try the sub-application approach for now.

**Reply by skrawcz** [reply]
@yiweig cool thanks. A follow up question or few:

1. how would you think about the "glue" code that needs to happen? e.g. how would you specify the subset of state you want each action to operate on (e.g. is it a list)? Or would you write to two separate places in state, one for each action?
2. What about reducing their results for a final state update?
3. Is it an all or nothing type action?

**Reply by yiweig** [reply]
Hey @skrawcz apologies for the late reply, I spent the past few days implementing the sub-application pattern for my use case and wanted to get it working first before responding to your (great) questions.

1. Our state is a dict, and we iterate over the `.items()` and dispatch each `(key, value)` pair to the sub-action.
2. For the final state reduction, I'm using a similar pattern from the example code that Elijah linked above: a single Action iterates over the dict items and creates sub-applications, which are then run using `joblib.Parallel()`. The results are collected into a single dict which is returned by that Action for the final state update.
3. For our case, I guess it would be an all or nothing type action, although it would be ok for us if one of the parallel calls errors out and does not return a response -- we would simply rely on the LLM outputs from the other parallel LLM calls.

Let me know if you would like me to clarify any of this!

**Reply by skrawcz** [reply]
@yiweig ICYMI we have a sketch of an API here - https://burr.dagworks.io/pull/370/concepts/parallelism/ would love any feedback and thoughts please :)

**Comment by skrawcz** [comment]
FYI [Parallelism V1](https://burr.dagworks.io/concepts/parallelism/) shipped with  shipped with `0.33.0`
